Task 6

iharmonyharper opened the pull request.

iharmonyharper committed:
Task 6

zverok commented on app/movie_collection.rb:
```diff
@@ -0,0 +1,76 @@
+        @movie_class = ModernMovie
+      when (2000...2100)
+        @movie_class = NewMovie
+      end
```
Во-первых, зачем `@movie_class` — instance variable, он же короткоживущий?
Во-вторых, в Ruby можно так (виднее, что все ветки вычисляют одну и ту же переменную):
```ruby
movie_class = case ....
  when 1900...1945
    AncientMovie
  when 1945...1968
    ClassicMovie
  ...
# или, для коротких выражений, так — некоторым больше нравится:
movie_class = case ....
  when 1900...1945 then AncientMovie
  when 1945...1968 then ClassicMovie
```

zverok replied:
И ещё один совет (необязательный к исполнению) — в принципе, соответствие «период → класс» похоже на неизменяемые данные — можно прикинуть, не станет ли лучше код, если выразить это не кодом, а данными (константой)

zverok commented on app/theater/base_theater.rb:
```diff
@@ -0,0 +1,32 @@
+  private
+
+  def random_movie(movies)
+    movies.sort_by(&:rating).last(TOP).sample(1)[0]
```
Таким образом у менее рейтинговых фильмов никаких шансов оказаться просмотренными, а речь шла о вероятностях же. Разумный подход для этого случая — `sort_by { |m| m.rating * rand }.last` — таким образом сортировка будет случайной, но со «сдвинутыми» вероятностями.
И, к слову, `sample(1)[0]` → `sample` ;)

zverok commented on app/theater/base_theater.rb:
```diff
@@ -0,0 +1,32 @@
+    @movies_collection = movies_collection
+  end
+
+  def show(_movie = nil, **filter)
```
Плохое название переменной — в Ruby есть соглашение что с `_` начинается название переменной, которая использоваться не будет (например, если мы переопределяем метод родителя и хотим сохранить тот же набор входных аргументов, но использовать их не собираемся)

zverok commented on app/theater/base_theater.rb:
```diff
@@ -0,0 +1,32 @@
+  def random_movie(movies)
+    movies.sort_by(&:rating).last(TOP).sample(1)[0]
+  end
+end
```
С `customer`-ом получилось не очень хорошо с архитектурной точки зрения. Получается, что у нас метод `show` имеет довольно странное «предусловие» — «пожалуйста, используйте `customer =`, передав объект конкретного класса `Customer`, перед каждым вызовом `show`!». Причём внутри `show` он сбрасывается, и даже если мы программируем что-то вроде «показать последовательно вот эти три фильма одному и тому же человеку», всё равно получается спагетти-код вида `theatre.customer = ...; theatre.show; theatre.customer = ...; theatre.show` и т.д.

Можно по-разному уточнить требования, как кинотеатры взаимодействуют с покупателями, но давайте сведём всё к принципу YAGNI (you ain't gonna need it) и последуем заданию — «сколько на счету денег» — это просто часть состояния Netflix-кинотеатра (и, кстати, только его!) — мы считаем что каждый объект-netflix соответствует ровно одному клиенту онлайн-кинотеатра и имеет дело только с ним. Иными словами, давайте сделаем меньше и проще, и избавимся от класса `Customer` (а можно и от `Account`)

zverok commented on app/theater/base_theater.rb:
```diff
@@ -0,0 +1,32 @@
+  end
+
+  def how_much?(movie_title)
+    movies_collection.filter(title: movie_title)[0].ticket_price
```
Дело личного вкуса, но как по мне, в длинных цепочках методов `.first` выглядит органичнее, чем разрывающий цепочку `[0]` — и более универсален (если какой-то метод вдруг вернёт `Enumerator`, а не `Array`, то `first` сработает, а `[0]` — нет)

zverok commented on app/theater/theater.rb:
```diff
@@ -0,0 +1,32 @@
+  def when?(movie_title)
+    movie = movies_collection.filter(title: movie_title)[0]
+    if %w[Drama Horror].any? { |genre| movie.genre.include? genre }
+      'evening'
```
Для таких целей лучше символы использовать, а не строки.

zverok commented on app/theater/theater.rb:
```diff
@@ -0,0 +1,32 @@
+    when 'evening'
+      { genre: /(drama)|(horror)/i }
+    end
+  end
```
Заметим, что одно и то же знание «такому-то периоду соответствуют такие-то фильтры» — закодировано независимо в двух разных методах, что нехорошо. Если требования чуток поменяются — надо не забыть поменять их в двух местах. Может, попробовать вынести это знание в константу, и использовать в обоих методах?

zverok commented on app/movie/ancient_movie.rb:
```diff
@@ -0,0 +1,14 @@
+
+class AncientMovie < Movie
+  def initialize(movies_collection: [], **args)
+    super(movies_collection: movies_collection, **args)
```
К слову, если в `super` передаётся тот же набор аргументов, что пришёл в текущий метод, можно писать просто `super` (это будет значить вызвать метод родителя с теми же аргументами; а чтобы вызвать без аргумента вообще — `super()`)

zverok commented on app/movie/modern_movie.rb:
```diff
@@ -0,0 +1,14 @@
+  def initialize(movies_collection: [], **args)
+    super(movies_collection: movies_collection, **args)
+    @period = :modern
+    @ticket_price = 3.00
```
Вот эти 2 строчки повторяются 4 раза, но нужны ли они?
Название периода можно однозначно вывести из названия класса, а цена билета — константа, а не переменная.

Commit:
Task 6*

zverok commented on app/movie_collection.rb:
```diff
@@ -0,0 +1,66 @@
+  def get_movies
+    @collection_raw_data.map do |data|
+      movie_class = Movie.movies_catalog.select { |_k, v| v[:year] === data[:year].to_i }.keys.first
+      Object.const_get(movie_class).new(movies_collection: self, **data)
```
Я бы сказал, что этой функциональности будет лучше в классе `Movie`, чисто по логике. Чтобы здесь осталось что-то вроде `Movie.build(self, **data)`

zverok commented on app/movie_collection.rb:
```diff
@@ -0,0 +1,66 @@
+
+  def get_movies
+    @collection_raw_data.map do |data|
+      movie_class = Movie.movies_catalog.select { |_k, v| v[:year] === data[:year].to_i }.keys.first
```
К слову, в таких, менее generic случаях, лучше использовать не `===`, а `cover?` — читателю не придётся задумываться что ж там слева, что к нему `===` применяется.

И ещё кстати: `select.first` → `detect`

zverok commented on app/movie/movie.rb:
```diff
@@ -0,0 +1,65 @@
+    @rating = @rating.to_f.round(2)
+    @year = @year.to_i
+    @period = Movie.movies_catalog[self.class.to_s][:period]
+    @ticket_price = Movie.movies_catalog[self.class.to_s][:ticket_price]
```
Думаю, что эти два вполне можно сделать методами — хранить их в переменных незачем (они же не _переменные_, т.е. изменяться не будут). А `period` так и вовсе из имени класса можно выводить, хранить его незачем.

zverok commented on app/theater/base_theater.rb:
```diff
@@ -0,0 +1,41 @@
+  end
+
+  def pay(amount)
+    raise(PaymentError, 'Invalid payment operation') if amount.to_f.round(2) < 0.00
```
Не стоит усложнять с округлением и сравнением с дробным нулём. С дробными числами всё сложно, конечно, но в данном случае и просто `amount < 0` сработало бы (что можно заменить в современном руби на `negative?`, но в данном случае не стоит, потому што на самом деле правильное тут условие, мне кажется, `<= 0`)

zverok commented on app/theater/base_theater.rb:
```diff
@@ -0,0 +1,41 @@
+class AccountBalanceError < StandardError
+end
+class PaymentError < StandardError
+end
```
Такие ошибки, используемые только одним классом, лучше в него же и вкладывать.

zverok commented on app/theater/netflix.rb:
```diff
@@ -0,0 +1,4 @@
+require_relative 'base_theater'
+
+class Netflix < BaseTheater
+end
```
Вообще, у нас тут получается архитектурная ошибка. Если Netflix — это в точности BaseTheater, то он не нужен вообще. Но по заданию у нетфликса есть функционал оплаты и баланса, а у обычного кинотеатра нету (в чём есть логика — объект нетфликса соответствует одному аккаунту одного пользователя в онлайн-кинотеатре, а объект theatre — зданию с кучей случайных посетителей, чей там «баланс»?)

zverok commented on app/theater/theater.rb:
```diff
@@ -0,0 +1,28 @@
+      :day
+    else
+      :never
+    end
```
А тут бы тоже константу `FILTERS` использовать, м? Для этого придётся немножко потюнить функционал фильтрации просто, чтобы можно было написать в фильтрах `genre: %w[Drama Horror]` — и тогда всё аккуратно сведётся к обобщённому коду.

zverok commented on spec/base_theater_spec.rb:
```diff
@@ -0,0 +1,47 @@
+
+  it 'is available as described_class' do
+    expect(base_theater.class).to eq(described_class)
+  end
```
а) в чём смысл этого теста?..
б) `it { is_expected.to be_a(described_class) }`

zverok commented on spec/base_theater_spec.rb:
```diff
@@ -0,0 +1,47 @@
+  it '#pay should add money to balance' do
+    base_theater.pay(1.00)
+    expect(base_theater.balance).to eq 1.00
+  end
```
тут есть полезный матчер `expect {}.to change`

zverok commented on spec/base_theater_spec.rb:
```diff
@@ -0,0 +1,47 @@
+  it '#show should withdraw money from balance' do
+    base_theater.pay(money)
+    base_theater.show(movie)
+    expect(base_theater.balance).to eq(money - movie.ticket_price)
```
а) опять же, `expect {}.to change` будет выразительнее
б) в тестах лучше использовать конкретные значения числовые (в смысле, справа, в матчерах), а не свойства тестируемых объектов — иначе может статься, что два бага прячут друг друга (например, цена кина не вычлась... но почему-то она в тесте равна 0, так что никто этого не заметил)

zverok commented on spec/base_theater_spec.rb:
```diff
@@ -0,0 +1,47 @@
+    expect(base_theater.balance).to eq(money - movie.ticket_price)
+  end
+
+  [0.00, 0, 0.01, -0, -0.00, 0.0000000001].each do |balance|
```
Не уверен, что такая детализация теста необходима :) В чём юз-кейс проверки? Тесты обработки ошибочного поведения — это хорошо, но они в идеале должны тестировать ошибки, которые правда можно совершить, а не любую странную ситуацию (а то можно дойти до того, что при тестировании калькулятора проверять, как он себя поведёт, если закончится место на диске, упадёт сеть, заглючит видеокарта или война начнётся).

Мы же здесь работаем с чем-то похожим на денежные суммы — и мне кажется, вероятностью того, что у кого-то на балансе окажется 4.9999999999999999, а кино стоит 5, можно пренебречь (вообще, во избежание мизерных глюков, на самом деле деньги в дробных переменных не держат, но мы этим займёмся ещё в одном из следующих уроков)

То есть одного теста с балансом меньше необходимого вполне достаточно.

zverok commented on spec/base_theater_spec.rb:
```diff
@@ -0,0 +1,47 @@
+    base_theater.pay(money)
+    filter = { genre: 'Comedy', period: :classic }
+    movie = base_theater.movies_collection.filter(filter).first
+    base_theater.show(filter)
```
Кстати, тут так и нет проверки, что он кино _показывает_ (напоминаю, что мы имитируем это, _печатая_ `Now showing: название фильма`)

zverok commented on spec/base_theater_spec.rb:
```diff
@@ -0,0 +1,47 @@
+    movie = base_theater.movies_collection.filter(filter).first
+    base_theater.show(filter)
+    expect(base_theater.balance).to eq(money - movie.ticket_price)
+  end
```
Посмотрите http://www.betterspecs.org/ (только обязательно на англ., русская версия устарела) — тут бы подошёл тестовый контекст, как-то так что ли:
```ruby
context '#show' do
  before { base_theater.pay(money) }
  let(:filter) { { genre: 'Comedy', period: :classic } }
  let(:movie) { base_theater.movies_collection.filter(filter).first }
  subject { base_theater.show(filter) }
  it 'shows movie' do
    expect { subject }.to ouput....
  end
end
```

Прелесть контекстов в том, что их легко обобщать (и в соседних тестах не приходится снова и снова писать `pay` — если вынести `before { theater.pay(money) }` на верхний уровень), и что они лучше показывают что у нас тут предусловия, а что — объект тестирования.

zverok commented on spec/netflix_spec.rb:
```diff
@@ -0,0 +1,30 @@
+  it 'is available as described_class' do
+    expect(netflix.class).to eq(described_class)
+    expect(netflix.class.superclass).to eq(base_theater.class)
+  end
```
опять же — смысл теста?..

zverok commented on spec/netflix_spec.rb:
```diff
@@ -0,0 +1,30 @@
+
+  it '#how_much?' do
+    expect(movie.ticket_price).to eq(ticket_price)
+    expect(netflix.how_much?('The Terminator')).to eq(movie.ticket_price)
```
Хорошо бы потестировать, что он сделает если фильм не найден

zverok commented on spec/shared_context_test_data.rb:
```diff
@@ -0,0 +1,12 @@
+  let(:theater) { Theater.new(movies_collection: movies) }
+  let(:base_theater) { BaseTheater.new(movies_collection: movies) }
+  let(:ticket_price) { 3.00 }
+  let(:money) { 25.00 }
```
На самом деле это не очень хороший shared context. 
У нас получается, что в значительной части тестов не видно «откуда вообще взялся этот объект» (потому что он создаётся вместе с кучей других, неинтересных объектов, в `context 'test data'`), или «а сколько у нас денег-то». Это один из тех случаев, когда читаемость важнее DRY. Попробуйте обойтись без него (раскидав необходимые переменные по тестам — их окажется, для каждого, не так-то и много)

zverok commented on spec/theater_spec.rb:
```diff
@@ -0,0 +1,33 @@
+      end
+      expect(possible_movies).to include(theater.show(time))
+    end
+  end
```
Тут тоже нужно бы `context` а не длинный тест.
И хорошо бы отдельный тест что оно сделает на незнакомый период.
И, кстати! В задании «передать время, в которое вы хотите посмотреть кино» — то есть или объект Time, или число с номером часа, или строку вида '11:30' (в смысле, выберите что-то одно из этого) — потому что зритель не обязан знать, когда с точки зрения директора кинотеатра начинается день, вечер и ночь.

iharmonyharper committed:
Task 6*

iharmonyharper committed:
Task 6**

Commit:
Task 6**
#when?

iharmonyharper committed:
Task 6**
#show

zverok commented on app/movie/movie.rb:
```diff
@@ -0,0 +1,88 @@
+
+  def initialize(movies_collection: [], **args)
+    args.each do |k, v|
+      value = v.split(/\s{0},\s{0}/) rescue [v]
```
а) `\s*` лучче
б) зачем `rescue`, что оно тут делает вообще?..

zverok commented on app/movie/movie.rb:
```diff
@@ -0,0 +1,88 @@
+  def initialize(movies_collection: [], **args)
+    args.each do |k, v|
+      value = v.split(/\s{0},\s{0}/) rescue [v]
+      self.class.send(:attr_reader, k)
```
Нехорошо это, мне кажется. 
Во-первых, с той точки зрения что мы 250 раз определяем классу одни и те же attr_readers, что _довольно странно_.
Во-вторых, получается (если думать о классе фильма как об отдельной полезной штуке), что можно сделать `Movie.new(foo: 1, bar: 2)` и он как бы «работает» (хотя на самом деле нет — большинство ожиданий от этого класса сломаются).

В таких случаях всё же лучше явно описать, какие `attr_reader`, даже если это кажется «недостаточно DRY»

zverok commented on app/movie/movie.rb:
```diff
@@ -0,0 +1,88 @@
+
+  def genres
+    [@genre].join(',')
+  end
```
Ээээ что это и зачем?..

zverok commented on app/movie/movie.rb:
```diff
@@ -0,0 +1,88 @@
+  end
+
+  def period
+    self.class.to_s.split('Movie').first.downcase.to_sym
```
а) лучше `.class.name` чем `.class.to_s` (более явно выражает намерение)
б) выделение нужной части тоже лучше не через `split` — например, можно так: `scan(/(\w+)Movie/).flatten.first`, или просто `sub('Movie', '')` — оба варианта лучше показывают что и откуда мы там выделяем.

zverok replied:
Этот коммент всё ещё актуален :(

zverok commented on app/theater/base_theater.rb:
```diff
@@ -0,0 +1,38 @@
+require_relative '../movie_collection'
+
+class MovieSearchError < StandardError
```
Я, кажется уже писал (или забыл) — такие «частные» (используемые только этим классом) ошибки лучше в сам же класс и вкладывать — и чтобы не засорять общее пространство имён, и чтобы было понятнее, откуда оно прилетело

zverok commented on app/theater/theater.rb:
```diff
@@ -0,0 +1,48 @@
+  NOT_DRAMA_HORROR_REGEX = /^(?!.*(drama|horror)).*$/i
+  NOT_ANCIENT_REGEX = /^(?!.*(ancient)).*$/i
+
+  FILTERS = { morning: { period: :ancient, genres: NOT_DRAMA_HORROR_REGEX },
```
Ага, я понял зачем вводился отдельный `genres`, но тут у меня два важных замечания:
1. вообще-то в задании ничего такого не было (про то что «оно не имеет права быть хоррором»), но вывод в принципе логичный;
2. мне кажется, такой подход очень путает здесь логику; лично я предпочёл бы дополнительный фильтр `exclude_genres:` — с ним придётся повозиться немножко (особенно если реализовать в обобщённом виде, `exclude_XXXX`), зато здесь, на более высоком уровне, будет сильно читабельнее.

zverok commented on app/theater/theater.rb:
```diff
@@ -0,0 +1,48 @@
+
+  SCHEDULE = { morning: ['09:00', '14:00'] ,
+    day:  ['14:00', '19:00'] ,
+    evening:  ['19:00', '23:00'] }
```
range тут будет лучше смотреться (особо никакой пользы не принесёт, но в декоративных целях — понятнее будет)

zverok commented on app/theater/theater.rb:
```diff
@@ -0,0 +1,48 @@
+    t = time_filters.detect{|(time, filter)|
+      movie = movies_collection.filter({title: movie_title}.merge(filter))
+      time unless movie.empty?
+    }
```
Ту же идею можно красивше:
```ruby
time, _ = time_filters.detect { |_, filter| movies_collection.filter({title: movie_title}.merge(filter)).any? }
return nil unless time
....
````

Но логичнее сделать следующее (чисто по постановке задачки):
1. найти кино
2. определить, под какой фильтр оно попадает
3. вернуть соотв. время

Псевдокод:
```ruby
movie = movies_collection.filter(title: title).first 
return ... unless movie
time, _ = time_filters.detect { |_, filter| movie.matches?(filter) }
return ... unless time
...
```

Это получается как бы длиннее и менее изобретательно, зато чётко показывает ход мысли.

Можно оставить и текущую реализацию, только сделать её чище и читабельнее.

zverok commented on app/theater/theater.rb:
```diff
@@ -0,0 +1,48 @@
+
+  def time_filters
+    FILTERS
+  end
```
Зачем эти два метода? Константы и так хороши, в эээ сыром виде.

zverok commented on app/theater/theater.rb:
```diff
@@ -0,0 +1,48 @@
+
+  def filters(time)
+    FILTERS[time]
+  end
```
Опять же — зачем? Чем вызывать в других местах `filters(time)` красивее/лаконичнее/эффективнее, чем `FILTERS[time]`?..

zverok commented on app/theater/theater.rb:
```diff
@@ -0,0 +1,48 @@
+  end
+
+  def show(time)
+    t = schedule.select{|k,v| v[0] == time }.keys.first
```
Почему `==`? Если у нас в 09:10 спрашивают что посмотреть — мы скажем что в этот момент фильмов нету?.. Но ведь по постановке-то они есть!

Commit:
Task 6**
#matches? #exclude?

Commit:
Task 6**
tests

zverok commented on app/movie/movie.rb:
```diff
@@ -0,0 +1,126 @@
+
+  def self.movies_catalog
+    MOVIES_CATALOG
+  end
```
А зачем этот метод?..

zverok commented on app/theater/base_theater.rb:
```diff
@@ -0,0 +1,40 @@
+             Time.parse(when?(movie.title))
+           rescue
+             when?(movie.title)
+           end
```
Во-первых, почему это в базовом классе? По условию задачи, у нас Theater имеет метод `when?`, а Netflix — нет, соответственно попытка его обобщить не очень уместна.
Во-вторых, rescue — это способ обработки _исключительных_ ситуаций (exception, you know: что-то пошло так, как программист совершенно не ожидал, и вот последний барьер), а здесь оно используется потому что `when?` реализовано сильно по-разному.
В-третьих, я так полагаю что метод show должен всегда брать `now` в качестве начала показа (окей, для настоящих кинотеатров это не совсем так, но задание упрощено чтобы не требовать «рассчитать сколько раз фильм можно показать фильм за утро, и назначить время сеансов»)

zverok commented on app/theater/theater.rb:
```diff
@@ -0,0 +1,43 @@
+        time_range.cover? time
+      rescue
+        nil
+      end
```
А здесь зачем `rescue`? Какую именно возможную исключительную ситуацию мы здесь ловим?..
(Не говоря уж о том, что `rescue` без указания типа ожидаемой ошибки для _большинства_ ситуаций слишком «жадный», может случайно поймать и замаскировать совершенно неожиданную ошибку)

zverok commented on spec/netflix_spec.rb:
```diff
@@ -0,0 +1,87 @@
+        "Now showing: #{title} #{time.strftime('%H:%M')} - < duration unknown >"
+      end
+    }
+  end
```
Посмотрите на betterspecs.org (только на английском, русская версия сильно устарела) — там описаны несколько современных подходов к rspec, чтобы спеки были читабельнее.
Главные замечания:
1. Стоит использовать `let` вместо instance variables
2. В спеках нежелательно использовать — вообще где бы то ни было — if. `if` как бы намекает нам, что тест может пойти несколькими разными путями, и это неприемлемо (при каждом прогоне тесты должны делать точно одно и то же)
3. Лучше использовать коммуникативные имена для переменных, то есть movie_a, movie_b → old_movie, new_movie или чёт такое

zverok commented on spec/netflix_spec.rb:
```diff
@@ -0,0 +1,87 @@
+                subject
+              rescue
+                nil
+              end
```
Неприемлемо в спеках. Либо тестируемый код бросает ошибку, и мы тестируем это, либо он работает, и мы тестируем его работу. Это как с `if`-ами, не стоит писать тест «может сломается, а может нет, как пойдёт!»

zverok commented on spec/netflix_spec.rb:
```diff
@@ -0,0 +1,87 @@
+              rescue
+                nil
+              end
+      end.not_to change { @netflix.balance }
```
Лучше использовать вот такую форму: `change(@netflix, :balance)` — у неё лучше документируемость (когда падает, то пишет «expected not to change netflix balance», и при запуске `rspec --format doc` более информативно описывает ожидание.

zverok commented on spec/netflix_spec.rb:
```diff
@@ -0,0 +1,87 @@
+    subject { @netflix.show(genre: 'Comedy', period: :classic) }
+
+    it 'shows movie' do
+      expect { print subject }.to output(@expected_show_output.call('Roman Holiday')).to_stdout
```
Эээ нет. Какой смысл тестировать рубишную стандартную функцию `print` — мы и так знаем, что она работает :)
То есть тест должен быть на то, что делает этот самый `show` (а по заданию, кстати, он как раз печатать должен, а не просто строку возвращать)

zverok commented on spec/netflix_spec.rb:
```diff
@@ -0,0 +1,87 @@
+    subject { @netflix.show(genre: 'Comedy') }
+    it 'shows random movie' do
+      expect([@expected_show_output.call('Roman Holiday'),
+              @expected_show_output.call('Roman Holiday (new)')]).to include(subject)
```
Нет, это совсем нехорошо :( То есть использовать результат вызова какого-то proc для теста — это нечитаемо, а тесты должны быть самодостаточны и чётко описывать что должно произойти. Здесь могло бы быть например так:
```ruby
expect(subject)
  .to eq('.........длинная строка с ожидаемым описанием фильма....')
  .or eq('........ второй вариант')
```
... или даже ДВА разных теста. Первый — как в целом строка форматируется, второй — что какой-то из этих двух фильмов выберется:
```ruby
expect(subject)
  .to match(/^Now showing (.+) и т.д., общий паттерн как оно должно выглядеть/)
  .and include('Roman Holiday')
  .or include('Roman Holiday (new)')
```

zverok reviewed (changes requested):
Это всё ещё не ко всему коду замечания, но принципы я тут показываю. Не стесняйтесь ругаться со мной в комментах, почте или чятике, если не согласны или не понятно.

Commit:
Task 6**
delete #when from Netflix class
delete #def self.movies_catalog
delete most of 'rescue'
delete 'if' from specs
delete prints

Commit:
Task 6**
clean

zverok commented on app/movie/movie.rb:
```diff
@@ -0,0 +1,122 @@
+    @country = args[:country]
+    @date = args[:date]
+    @genre = begin
+               args[:genre].split(/\s{0},\s{0}/)
```
`\s{0}` → `\s*` (так тоже нормально, но лучше использовать максимально стандартные варианты, как минимум для простых регекспов)

zverok commented on app/movie/movie.rb:
```diff
@@ -0,0 +1,122 @@
+    @date = args[:date]
+    @genre = begin
+               args[:genre].split(/\s{0},\s{0}/)
+             rescue
```
А от чего защищает этот `rescue`? Какая конкретно ошибка может произойти в момент `split`?..

zverok commented on app/theater/base_theater.rb:
```diff
@@ -0,0 +1,36 @@
+    raise(MovieSearchError, "No results for '#{filter}'") unless movie
+    yield(movie) if block_given?
+    time = Time.now if /netflix/i === self.class.name
+    time = Time.parse(when?(movie.title)) if /theater/i === self.class.name
```
Вот эти две строчки — нарушение всех принципов наследования :) У нас же любой метод можно в классе-наследнике переопределить. То есть здесь должно быть что-то вроде `start_time(movie)` в `Netflix` определённое одним образом, а в `Theatre` — другим. Наследование ж для того и нужно, чтобы при одинаковом интерфейсе иметь разное поведение наследников.

zverok commented on spec/netflix_spec.rb:
```diff
@@ -0,0 +1,77 @@
+             subject
+           rescue
+             Netflix::AccountBalanceError
+           end
```
Ага, я осознал зачем тут `rescue`. Для таких случаев — хорошо объединять тесты с `.and` — получается читабельнее:
```ruby
expect { subject }
  .to raise_error(...)
  .and not_change(netflix, :balance)
```
Первый матчер проверяет ошибку и «проглатывает» её, второй тестирует отсутствие изменений (негативные матчеры делаются с помощью `RSpec::Matchers.define_negated_matcher`)

Код при этом получается максимально отражающим смысл и наши ожидания.

zverok reviewed (changes requested):
Тесты уже прям хороши стали! Особенно для `Theater`
Продолжаем :)

iharmonyharper committed:
Task 6**
-\s{0} → \s*
-delete rescue in genres and actors
-#period
-#start_time
-.and avoid_changing(netflix, :balance)

Commit:
Task 6**
 spec

Commit:
.travis.yml

Commit:
try travis

Commit:
try travis

Commit:
try travis

Commit:
try travis

Commit:
try travis

Commit:
try travis

Commit:
try travis

Commit:
try travis

Commit:
try travis

Commit:
try travis

Commit:
Merge branch 'master' into develop
